tree/ntupleutil/v7/src/RNTupleImporter.cxx

```diff
@@ -13,16 +13,331 @@
  * For the list of contributors see $ROOTSYS/README/CREDITS.             *
  *************************************************************************/
 
+#include <ROOT/RError.hxx>
+#include <ROOT/RField.hxx>
+#include <ROOT/RNTuple.hxx>
 #include <ROOT/RNTupleImporter.hxx>
+#include <ROOT/RNTupleOptions.hxx>
 #include <ROOT/RNTupleUtil.hxx>
+#include <ROOT/RPageStorage.hxx>
+#include <ROOT/RPageStorageFile.hxx>
 #include <ROOT/RStringView.hxx>
 
-#include <memory>
+#include <TBranch.h>
+#include <TClass.h>
+#include <TDataType.h>
+#include <TLeaf.h>
+#include <TLeafC.h>
+#include <TLeafElement.h>
+#include <TLeafObject.h>
 
-std::unique_ptr<ROOT::Experimental::RNTupleImporter>
-ROOT::Experimental::RNTupleImporter::Create(std::string_view /* sourceFile */, std::string_view /* treeName */,
-                                            std::string_view /* destFile */)
+#include <cassert>
+#include <cstdint>
+#include <cstring>
+#include <iostream>
+#include <utility>
+
+namespace {
+
+class RDefaultProgressCallback : public ROOT::Experimental::RNTupleImporter::RProgressCallback {
+private:
+   static constexpr std::uint64_t gUpdateFrequencyBytes = 50 * 1000 * 1000; // report every 50MB
+   std::uint64_t fNbytesNext = gUpdateFrequencyBytes;
+
+public:
+   virtual ~RDefaultProgressCallback() {}
+   void Call(std::uint64_t nbytesWritten, std::uint64_t neventsWritten) final
+   {
+      // Report if more than 50MB (compressed) where written since the last status update
+      if (nbytesWritten < fNbytesNext)
+         return;
+      std::cout << "Wrote " << nbytesWritten / 1000 / 1000 << "MB, " << neventsWritten << " entries" << std::endl;
+      fNbytesNext += gUpdateFrequencyBytes;
+   }
+
+   void Finish(std::uint64_t nbytesWritten, std::uint64_t neventsWritten) final
+   {
+      std::cout << "Done, wrote " << nbytesWritten / 1000 / 1000 << "MB, " << neventsWritten << " entries" << std::endl;
+   }
+};
+
+} // anonymous namespace
+
+ROOT::Experimental::RResult<void>
+ROOT::Experimental::RNTupleImporter::RCStringTransformation::Transform(const RImportBranch &branch, RImportField &field)
+{
+   *reinterpret_cast<std::string *>(field.fFieldBuffer) = reinterpret_cast<const char *>(branch.fBranchBuffer.get());
+   return RResult<void>::Success();
+}
+
+ROOT::Experimental::RResult<void>
+ROOT::Experimental::RNTupleImporter::RLeafArrayTransformation::Transform(const RImportBranch &branch,
+                                                                         RImportField &field)
+{
+   auto valueSize = field.fField->GetValueSize();
+   memcpy(field.fFieldBuffer, branch.fBranchBuffer.get() + (fNum * valueSize), valueSize);
```

Comment:
I'm wondering if there's any chance of avoiding the copy in this case.  :thinking:

Reply:
That should be the goal. I'm afraid at the moment the RNTuple untyped collection API limits us. Something to work on next year.